## data preprocessing [ML pipeline]
- Data collection -  web scrapping / database (sql) / data lake (Biquerry) -big data tools
- Data read -  csv,json,xl,.html

-----
## Data cleaning 

    - Duplicates 
        - Repated entry / contain same value or record 
        - Why we have to remove ?
            - Same data records can be seen in both train and test dataset result in data leakage
            when the are coming in data leakege
              . Feature scalling
              . Data splitting  
              . Model Training (ml models) 
              . Model Evaluvation (metrics) 
------   
    - Regex Handling -----> converted np.nan
    - Null value / NaN values (des stat, ml )
        - shows only the empty cell count using isnull() or isna() 
        - regex is not counted as null or NaN
        - How to handle Null or NaN ?
           1. Remove :
               - when your null value is low in count 
               - when cell are empty by mistake rows or column
               -  How?
                    - Drop the unwanted columns or rows
           2. Replace : 
               - when your null value is large in count 
               - How ?
                   - Median - because it is not affected by outliers
                   - Imputations :
                       - A,B,C,D,E,F,J
                       -J HAS a null value in row 
                       data split
                       - J HAS a some null value in row
                       - J HAS a no null value in row 
                       model training
                       - x(remaining column s rows)and y is ---J HAS a no null value in row 
                       - model - Training - X and Y 
                       model predict 
                       - x(remaining column s rows)and y is-----J HAS a some null value in row
                       - model - Predicting - X -----> model gives predicted Y
       
---
    - outlier (wrong / extreme data) --- Normal distribution (Emprical formula)
    
-------
- Feature encoding 
    - Categorical to number 
- Feature Extraction (NLP,image dataset)
- Feature selection (Inferencial statistics, models)
- Feature scalling (normal distribution/ standard normal ( z-test)) 
- Data splitting  
- Model Training (ml models) 
- Model Evaluvation (metrics) 
- Model Hyperparameter tunning 
- Model save (.pkl,joblib)------- using this i can go fro predict Predict 

--------------------------

In [5]:
import pandas as pd
import numpy as np
import sys

In [6]:
# increase the default column
pd.set_option('display.max_rows', 100)
pd.set_option('display.max_columns', 100)

In [7]:
df=pd.read_csv("D:\hypothyroid.csv")

In [8]:
df

age sex on thyroxine query on thyroxine on antithyroid medication sick  \
0     41   F            f                  f                         f    f   
1     23   F            f                  f                         f    f   
2     46   M            f                  f                         f    f   
3     70   F            t                  f                         f    f   
4     70   F            f                  f                         f    f   
...   ..  ..          ...                ...                       ...  ...   
3767  30   F            f                  f                         f    f   
3768  68   F            f                  f                         f    f   
3769  74   F            f                  f                         f    f   
3770  72   M            f                  f                         f    f   
3771  64   F            f                  f                         f    f   

     pregnant thyroid surgery I131 treatment query hypothyroid  \
0           f               f              f                 f   
1           f               f              f                 f   
2           f               f              f                 f   
3           f               f              f                 f   
4           f               f              f                 f   
...       ...             ...            ...               ...   
3767        f               f              f                 f   
3768        f               f              f                 f   
3769        f               f              f                 f   
3770        f               f              f                 f   
3771        f               f              f                 f   

     query hyperthyroid lithium goitre tumor hypopituitary psych TSH measured  \
0                     f       f      f     f             f     f            t   
1                     f       f      f     f             f     f            t   
2                     f       f      f     f             f     f            t   
3                     f       f      f     f             f     f            t   
4                     f       f      f     f             f     f            t   
...                 ...     ...    ...   ...           ...   ...          ...   
3767                  f       f      f     t             f     f            f   
3768                  f       f      f     f             f     f            t   
3769                  t       f      f     f             f     f            t   
3770                  f       f      f     f             f     f            t   
3771                  f       f      f     f             f     f            t   

       TSH T3 measured   T3 TT4 measured  TT4 T4U measured   T4U FTI measured  \
0      1.3           t  2.5            t  125            t  1.14            t   
1      4.1           t    2            t  102            f     ?            f   
2     0.98           f    ?            t  109            t  0.91            t   
3     0.16           t  1.9            t  175            f     ?            f   
4     0.72           t  1.2            t   61            t  0.87            t   
...    ...         ...  ...          ...  ...          ...   ...          ...   
3767     ?           f    ?            f    ?            f     ?            f   
3768   NaN           t  2.1            t  124            t  1.08            t   
3769   NaN           t  1.8            t  112            t  1.07            t   
3770   NaN           t    2            t   82            t  0.94            t   
3771   NaN           t  2.2            t   99            t  1.07            t   

      FTI TBG measured TBG referral source binaryClass  
0     109            f   ?            SVHC           P  
1       ?            f   ?           other           P  
2     120            f   ?           other           P  
3       ?            f   ?           other           P  
4      70            f   ?      

In [9]:
df.tail()

age sex on thyroxine query on thyroxine on antithyroid medication sick  \
3767  30   F            f                  f                         f    f   
3768  68   F            f                  f                         f    f   
3769  74   F            f                  f                         f    f   
3770  72   M            f                  f                         f    f   
3771  64   F            f                  f                         f    f   

     pregnant thyroid surgery I131 treatment query hypothyroid  \
3767        f               f              f                 f   
3768        f               f              f                 f   
3769        f               f              f                 f   
3770        f               f              f                 f   
3771        f               f              f                 f   

     query hyperthyroid lithium goitre tumor hypopituitary psych TSH measured  \
3767                  f       f      f     t             f     f            f   
3768                  f       f      f     f             f     f            t   
3769                  t       f      f     f             f     f            t   
3770                  f       f      f     f             f     f            t   
3771                  f       f      f     f             f     f            t   

      TSH T3 measured   T3 TT4 measured  TT4 T4U measured   T4U FTI measured  \
3767    ?           f    ?            f    ?            f     ?            f   
3768  NaN           t  2.1            t  124            t  1.08            t   
3769  NaN           t  1.8            t  112            t  1.07            t   
3770  NaN           t    2            t   82            t  0.94            t   
3771  NaN           t  2.2            t   99            t  1.07            t   

      FTI TBG measured TBG referral source binaryClass  
3767    ?            f   ?           other           P  
3768  114            f   ?             SVI           P  
3769  105            f   ?           other           P  
3770   87            f   ?             SVI           P  
3771   92            f   ?           other           P

In [10]:
df.shape

(3772, 30)

In [11]:
df=df.drop_duplicates()
df

age sex on thyroxine query on thyroxine on antithyroid medication sick  \
0     41   F            f                  f                         f    f   
1     23   F            f                  f                         f    f   
2     46   M            f                  f                         f    f   
3     70   F            t                  f                         f    f   
4     70   F            f                  f                         f    f   
...   ..  ..          ...                ...                       ...  ...   
3767  30   F            f                  f                         f    f   
3768  68   F            f                  f                         f    f   
3769  74   F            f                  f                         f    f   
3770  72   M            f                  f                         f    f   
3771  64   F            f                  f                         f    f   

     pregnant thyroid surgery I131 treatment query hypothyroid  \
0           f               f              f                 f   
1           f               f              f                 f   
2           f               f              f                 f   
3           f               f              f                 f   
4           f               f              f                 f   
...       ...             ...            ...               ...   
3767        f               f              f                 f   
3768        f               f              f                 f   
3769        f               f              f                 f   
3770        f               f              f                 f   
3771        f               f              f                 f   

     query hyperthyroid lithium goitre tumor hypopituitary psych TSH measured  \
0                     f       f      f     f             f     f            t   
1                     f       f      f     f             f     f            t   
2                     f       f      f     f             f     f            t   
3                     f       f      f     f             f     f            t   
4                     f       f      f     f             f     f            t   
...                 ...     ...    ...   ...           ...   ...          ...   
3767                  f       f      f     t             f     f            f   
3768                  f       f      f     f             f     f            t   
3769                  t       f      f     f             f     f            t   
3770                  f       f      f     f             f     f            t   
3771                  f       f      f     f             f     f            t   

       TSH T3 measured   T3 TT4 measured  TT4 T4U measured   T4U FTI measured  \
0      1.3           t  2.5            t  125            t  1.14            t   
1      4.1           t    2            t  102            f     ?            f   
2     0.98           f    ?            t  109            t  0.91            t   
3     0.16           t  1.9            t  175            f     ?            f   
4     0.72           t  1.2            t   61            t  0.87            t   
...    ...         ...  ...          ...  ...          ...   ...          ...   
3767     ?           f    ?            f    ?            f     ?            f   
3768   NaN           t  2.1            t  124            t  1.08            t   
3769   NaN           t  1.8            t  112            t  1.07            t   
3770   NaN           t    2            t   82            t  0.94            t   
3771   NaN           t  2.2            t   99            t  1.07            t   

      FTI TBG measured TBG referral source binaryClass  
0     109            f   ?            SVHC           P  
1       ?            f   ?           other           P  
2     120            f   ?           other           P  
3       ?            f   ?           other           P  
4      70            f   ?      

In [140]:
df.shape

(3711, 30)

In [141]:
df.describe()

age   sex on thyroxine query on thyroxine on antithyroid medication  \
count   3711  3711         3711               3711                      3711   
unique    94     3            2                  2                         2   
top       59     F            f                  f                         f   
freq      95  2424         3247               3661                      3669   

        sick pregnant thyroid surgery I131 treatment query hypothyroid  \
count   3711     3711            3711           3711              3711   
unique     2        2               2              2                 2   
top        f        f               f              f                 f   
freq    3564     3658            3658           3652              3477   

       query hyperthyroid lithium goitre tumor hypopituitary psych  \
count                3711    3711   3711  3711          3711  3711   
unique                  2       2      2     2             2     2   
top                     f       f      f     f             f     f   
freq                 3477    3693   3677  3615          3710  3527   

       TSH measured   TSH T3 measured    T3 TT4 measured   TT4 T4U measured  \
count          3711  3707        3711  3711         3711  3711         3711   
unique            2   288           2    70            2   242            2   
top               t     ?           t     ?            t     ?            t   
freq           3402   309        3002   709         3540   171         3384   

         T4U FTI measured   FTI TBG measured   TBG referral source binaryClass  
count   3711         3711  3711         3711  3711            3711        3711  
unique   147            2   235            1     1               5           2  
top        ?            t     ?            f     ?           other           P  
freq     327         3386   325         3711  3711            2140        3420

## This is an imbalance data set 
     . set column contain f 2424 out of 3711 row 
     . Binary class column contain negative 3420 out of 3711 row
     . so it is an  imbalance data set 


In [142]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3711 entries, 0 to 3771
Data columns (total 30 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   age                        3711 non-null   object
 1   sex                        3711 non-null   object
 2   on thyroxine               3711 non-null   object
 3   query on thyroxine         3711 non-null   object
 4   on antithyroid medication  3711 non-null   object
 5   sick                       3711 non-null   object
 6   pregnant                   3711 non-null   object
 7   thyroid surgery            3711 non-null   object
 8   I131 treatment             3711 non-null   object
 9   query hypothyroid          3711 non-null   object
 10  query hyperthyroid         3711 non-null   object
 11  lithium                    3711 non-null   object
 12  goitre                     3711 non-null   object
 13  tumor                      3711 non-null   object
 14  hypopitu

In [143]:
# how to find unwanted entity like , *$###^%
df.nunique()

age                           94
sex                            3
on thyroxine                   2
query on thyroxine             2
on antithyroid medication      2
sick                           2
pregnant                       2
thyroid surgery                2
I131 treatment                 2
query hypothyroid              2
query hyperthyroid             2
lithium                        2
goitre                         2
tumor                          2
hypopituitary                  2
psych                          2
TSH measured                   2
TSH                          288
T3 measured                    2
T3                            70
TT4 measured                   2
TT4                          242
T4U measured                   2
T4U                          147
FTI measured                   2
FTI                          235
TBG measured                   1
TBG                            1
referral source                5
binaryClass                    2
dtype: int

In [144]:
df['sex'].unique()

array(['F', 'M', '?'], dtype=object)

In [145]:
df['age'].unique()

array(['41', '23', '46', '70', '18', '59', '80', '66', '68', '84', '67',
       '71', '28', '65', '42', '63', '51', '81', '54', '55', '60', '25',
       '73', '34', '78', '37', '85', '26', '58', '64', '44', '48', '61',
       '35', '83', '21', '87', '53', '77', '27', '69', '74', '38', '76',
       '45', '36', '22', '43', '72', '82', '31', '39', '49', '62', '57',
       '1', '50', '30', '29', '75', '19', '7', '79', '17', '24', '15',
       '32', '47', '16', '52', '33', '13', '10', '89', '56', '20', '90',
       '40', '88', '14', '86', '94', '12', '4', '11', '8', '5', '455',
       '2', '91', '6', '?', '93', '92'], dtype=object)

In [146]:
df['binaryClass'].unique()

array(['P', 'N'], dtype=object)

In [147]:
# all unique value in data  frame using for loop 
for i in df.columns:
    print(i)
    print(df[i].unique())
    print('--------------------')

age
['41' '23' '46' '70' '18' '59' '80' '66' '68' '84' '67' '71' '28' '65'
 '42' '63' '51' '81' '54' '55' '60' '25' '73' '34' '78' '37' '85' '26'
 '58' '64' '44' '48' '61' '35' '83' '21' '87' '53' '77' '27' '69' '74'
 '38' '76' '45' '36' '22' '43' '72' '82' '31' '39' '49' '62' '57' '1' '50'
 '30' '29' '75' '19' '7' '79' '17' '24' '15' '32' '47' '16' '52' '33' '13'
 '10' '89' '56' '20' '90' '40' '88' '14' '86' '94' '12' '4' '11' '8' '5'
 '455' '2' '91' '6' '?' '93' '92']
--------------------
sex
['F' 'M' '?']
--------------------
on thyroxine
['f' 't']
--------------------
query on thyroxine
['f' 't']
--------------------
on antithyroid medication
['f' 't']
--------------------
sick
['f' 't']
--------------------
pregnant
['f' 't']
--------------------
thyroid surgery
['f' 't']
--------------------
I131 treatment
['f' 't']
--------------------
query hypothyroid
['f' 't']
--------------------
query hyperthyroid
['f' 't']
--------------------
lithium
['f' 't']
--------------------
goitre


In [148]:
df.isnull().count()

age                          3711
sex                          3711
on thyroxine                 3711
query on thyroxine           3711
on antithyroid medication    3711
sick                         3711
pregnant                     3711
thyroid surgery              3711
I131 treatment               3711
query hypothyroid            3711
query hyperthyroid           3711
lithium                      3711
goitre                       3711
tumor                        3711
hypopituitary                3711
psych                        3711
TSH measured                 3711
TSH                          3711
T3 measured                  3711
T3                           3711
TT4 measured                 3711
TT4                          3711
T4U measured                 3711
T4U                          3711
FTI measured                 3711
FTI                          3711
TBG measured                 3711
TBG                          3711
referral source              3711
binaryClass   

In [149]:
df.isnull().sum() 

age                          0
sex                          0
on thyroxine                 0
query on thyroxine           0
on antithyroid medication    0
sick                         0
pregnant                     0
thyroid surgery              0
I131 treatment               0
query hypothyroid            0
query hyperthyroid           0
lithium                      0
goitre                       0
tumor                        0
hypopituitary                0
psych                        0
TSH measured                 0
TSH                          4
T3 measured                  0
T3                           0
TT4 measured                 0
TT4                          0
T4U measured                 0
T4U                          0
FTI measured                 0
FTI                          0
TBG measured                 0
TBG                          0
referral source              0
binaryClass                  0
dtype: int64

In [151]:
df['TSH'].unique()

array(['1.3', '4.1', '0.98', '0.16', '0.72', '0.03', '?', '2.2', '0.6',
       '2.4', '1.1', '2.8', '3.3', '12', '1.2', '1.5', '6', '2.1', '0.1',
       '0.8', '1.9', '3.1', '0.2', '13', '0.3', '0.035', '2.5', '0.5',
       '1.7', '7.3', '1.8', '0.26', '45', '5.4', '0.99', '0.25', '0.92',
       '0.15', '0.64', '1', '0.4', '2', '2.6', '14.8', '15', '19', '0.02',
       '3', '2.9', '3.2', '9', '1.6', '4.3', '0.005', '0.31', '0.61',
       '0.05', '7.8', '160', '0.025', '1.4', '0.01', '8.8', '151', '0.04',
       '3.9', '9.4', '2.7', '2.3', '0.94', '0.045', '3.5', '0.88', '0.08',
       '4.5', '0.68', '0.7', '0.67', '27', '6.1', '0.75', '0.55', '26',
       '5.2', '0.77', '0.07', '0.9', '11.4', '143', '0.45', '0.57',
       '0.65', '0.015', '16', '108', '0.83', '9.2', '86', '0.62', '0.59',
       '9.1', '5.9', '52', '0.33', '31', '5.8', '0.28', '51', '6.3',
       '4.4', '9.6', '3.4', '0.09', '24', '0.76', '42', '25', '10', '4.6',
       '8.6', '0.66', '6.2', '0.79', '28', '0.86', '9.7',

In [153]:
df['TSH']=df['TSH'].fillna('m')

C:\Users\dhana\AppData\Local\Temp\ipykernel_15484\1416136108.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['TSH']=df['TSH'].fillna('m')


In [154]:
df['TSH'].unique()

array(['1.3', '4.1', '0.98', '0.16', '0.72', '0.03', '?', '2.2', '0.6',
       '2.4', '1.1', '2.8', '3.3', '12', '1.2', '1.5', '6', '2.1', '0.1',
       '0.8', '1.9', '3.1', '0.2', '13', '0.3', '0.035', '2.5', '0.5',
       '1.7', '7.3', '1.8', '0.26', '45', '5.4', '0.99', '0.25', '0.92',
       '0.15', '0.64', '1', '0.4', '2', '2.6', '14.8', '15', '19', '0.02',
       '3', '2.9', '3.2', '9', '1.6', '4.3', '0.005', '0.31', '0.61',
       '0.05', '7.8', '160', '0.025', '1.4', '0.01', '8.8', '151', '0.04',
       '3.9', '9.4', '2.7', '2.3', '0.94', '0.045', '3.5', '0.88', '0.08',
       '4.5', '0.68', '0.7', '0.67', '27', '6.1', '0.75', '0.55', '26',
       '5.2', '0.77', '0.07', '0.9', '11.4', '143', '0.45', '0.57',
       '0.65', '0.015', '16', '108', '0.83', '9.2', '86', '0.62', '0.59',
       '9.1', '5.9', '52', '0.33', '31', '5.8', '0.28', '51', '6.3',
       '4.4', '9.6', '3.4', '0.09', '24', '0.76', '42', '25', '10', '4.6',
       '8.6', '0.66', '6.2', '0.79', '28', '0.86', '9.7',

In [129]:
df=df.dropna()

In [76]:
df

age sex on thyroxine query on thyroxine on antithyroid medication sick  \
0     41   F            f                  f                         f    f   
1     23   F            f                  f                         f    f   
2     46   M            f                  f                         f    f   
3     70   F            t                  f                         f    f   
4     70   F            f                  f                         f    f   
...   ..  ..          ...                ...                       ...  ...   
3761  38   F            f                  f                         f    f   
3762  61   M            f                  f                         f    f   
3764  54   M            f                  f                         f    f   
3765  73   F            t                  f                         f    f   
3766  19   F            f                  f                         f    f   

     pregnant thyroid surgery I131 treatment query hypothyroid  \
0           f               f              f                 f   
1           f               f              f                 f   
2           f               f              f                 f   
3           f               f              f                 f   
4           f               f              f                 f   
...       ...             ...            ...               ...   
3761        f               f              f                 f   
3762        f               f              f                 f   
3764        f               f              f                 f   
3765        f               f              f                 f   
3766        f               f              f                 f   

     query hyperthyroid lithium goitre tumor hypopituitary psych TSH measured  \
0                     f       f      f     f             f     f            t   
1                     f       f      f     f             f     f            t   
2                     f       f      f     f             f     f            t   
3                     f       f      f     f             f     f            t   
4                     f       f      f     f             f     f            t   
...                 ...     ...    ...   ...           ...   ...          ...   
3761                  f       f      f     f             f     f            t   
3762                  f       f      f     f             f     f            f   
3764                  f       f      f     f             f     f            t   
3765                  f       f      f     f             f     f            t   
3766                  f       f      f     f             f     f            t   

       TSH T3 measured   T3 TT4 measured  TT4 T4U measured   T4U FTI measured  \
0      1.3           t  2.5            t  125            t  1.14            t   
1      4.1           t    2            t  102            f     ?            f   
2     0.98           f    ?            t  109            t  0.91            t   
3     0.16           t  1.9            t  175            f     ?            f   
4     0.72           t  1.2            t   61            t  0.87            t   
...    ...         ...  ...          ...  ...          ...   ...          ...   
3761   2.8           t  2.4            t   96            t  0.98            t   
3762     ?           t    1            t   72            t   0.7            t   
3764   1.2           t  1.9            t   89            t  0.85            t   
3765   8.5           t  2.1            t  104            t  1.13            t   
3766   8.8           t  2.7            t  108            t  1.11            t   

      FTI TBG measured TBG referral source binaryClass  
0     109            f   ?            SVHC           P  
1       ?            f   ?           other           P  
2     120            f   ?           other           P  
3       ?            f   ?           other           P  
4      70            f   ?      

In [77]:
df.isnull().sum()

age                          0
sex                          0
on thyroxine                 0
query on thyroxine           0
on antithyroid medication    0
sick                         0
pregnant                     0
thyroid surgery              0
I131 treatment               0
query hypothyroid            0
query hyperthyroid           0
lithium                      0
goitre                       0
tumor                        0
hypopituitary                0
psych                        0
TSH measured                 0
TSH                          0
T3 measured                  0
T3                           0
TT4 measured                 0
TT4                          0
T4U measured                 0
T4U                          0
FTI measured                 0
FTI                          0
TBG measured                 0
TBG                          0
referral source              0
binaryClass                  0
dtype: int64

In [ ]:
?
?
?
?
?
